src/config.rs

```diff
@@ -101,69 +101,118 @@ pub struct Provider {
+
+        self.public_url = Some(toml_config.server.public_url);
+
+        self.token_ttl = toml_config.crypto.token_ttl;
+
         for path in toml_config.crypto.keyfiles.iter() {
-            keys.push(try!(crypto::NamedKey::from_file(path)));
+            self.keyfiles.push(path.clone());
         }
 
-        let store = try!(store::Store::new(
-            &toml_config.redis.url,
-            toml_config.redis.session_ttl as usize,
-            toml_config.redis.cache_ttl as usize,
-            toml_config.redis.cache_max_doc_size as u64
-        ));
-
-        let mut providers: HashMap<String, Provider> = HashMap::new();
-        for (domain, settings) in toml_config.providers.iter() {
-            let provider = Provider {
-                client_id: settings.client_id.clone(),
-                secret: settings.secret.clone(),
-                discovery_url: settings.discovery_url.clone(),
-                issuer_domain: settings.issuer_domain.clone(),
-            };
-
-            providers.insert(domain.to_string(), provider);
-        }
+        self.redis_url = Some(toml_config.redis.url);
+
+        self.redis_session_ttl = toml_config.redis.session_ttl;
+
+        self.redis_cache_ttl = toml_config.redis.cache_ttl;
+
+        self.redis_cache_max_doc_size = toml_config.redis.cache_max_doc_size;
+
+        self.from_name = toml_config.smtp.from_name;
 
+        self.from_address = Some(toml_config.smtp.from_address);
 
-        Ok(Config {
-            listen_ip: toml_config.server.listen_ip,
-            listen_port: toml_config.server.listen_port,
-            public_url: toml_config.server.public_url,
-            token_ttl: toml_config.crypto.token_ttl,
+        self.smtp_server = Some(toml_config.smtp.server);
+
+        self
+    }
+
+    pub fn done(self) -> Config {
+        let keys = self.keyfiles.into_iter().filter_map(|path| {
+            crypto::NamedKey::from_file(&path).ok()
+        }).collect();
+
+        let store = store::Store::new(
+            &self.redis_url.unwrap(),
+            self.redis_cache_ttl as usize,
+            self.redis_session_ttl as usize,
+            self.redis_cache_max_doc_size as u64,
+        ).unwrap();
+
+        Config {
+            listen_ip: self.listen_ip,
+            listen_port: self.listen_port,
+            public_url: self.public_url.unwrap(),
+            token_ttl: self.token_ttl,
             keys: keys,
             store: store,
-            from_name: toml_config.smtp.from_name,
-            from_address: toml_config.smtp.from_address,
-            smtp_server: toml_config.smtp.server,
-            providers: providers,
+            from_name: self.from_name,
+            from_address: self.from_address.unwrap(),
+            smtp_server: self.smtp_server.unwrap(),
+            providers: self.providers,
             templates: Templates::default(),
-        })
+        }
     }
 }
+
+
+pub struct Config {
```

Comment:
This should be above Builder, so Config structs are grouped together. I think it'll also make more sense in the diff, then.